lib/mllp/client.ex

```diff
@@ -356,282 +381,384 @@ defmodule MLLP.Client do
-        err
+  def receiving({:call, _from} = event_kind, request, data)
+      when request in [:is_connected, :reconnect] do
+    connected(event_kind, request, data)
+  end
+
+  def receiving(event, unknown, _data) do
+    unexpected_message(:receiving, event, unknown)
+  end
+
+  ########################################
+  ### End of :gen_statem callbacks ###
+  ########################################
+
+  defp unexpected_message(state, event, message) do
+    Logger.warn(
+      "Event: #{inspect(event)} in state #{state}. Unknown message received => #{inspect(message)}"
+    )
+
+    :keep_state_and_data
+  end
+
+  ## Handle the (fragmented) responses to `send` request from a caller
+
+  defp handle_received(_reply, %{caller: nil} = data) do
+    ## No caller, ignore
+    data
+  end
+
+  defp handle_received(<<@header, _rest::binary>> = reply, data) do
+    receive_impl(reply, data)
+  end
+
+  ## No header in the first packet
+  defp handle_received(_reply, %{receive_buffer: []} = data) do
+    reply_to_caller({:error, :invalid_reply}, data)
+  end
+
+  ## The rest of MLLP (after the header was received)
+  defp handle_received(reply, data) do
+    receive_impl(reply, data)
+  end
+
+  def receive_impl(reply, %{receive_buffer: buffer, last_byte_received: last_byte} = data) do
+    new_buf = update_receive_buffer(buffer, reply)
+
+    case trailer_check(reply, last_byte) do
+      :data_after_trailer ->
+        Logger.error("Client #{inspect(self())} received data following the trailer")
+        reply_to_caller({:error, :data_after_trailer}, data)
+
+      true ->
+        Logger.debug("Client #{inspect(self())} received a full MLLP!")
+        reply_to_caller({:ok, IO.iodata_to_binary(new_buf)}, data)
+
+      false ->
+        Logger.debug("Client #{inspect(self())} received a MLLP fragment: #{reply}")
+
+        data
+        |> Map.put(:receive_buffer, new_buf)
+        |> Map.put(:last_byte_received, String.last(reply))
     end
   end
 
-  defp do_recv(state, length, :infinity) do
-    res = state.tcp.recv(state.socket, length, :infinity)
-    {res, 0}
+  ## "Split trailer" case
+  ##
+
+  defp trailer_check(<<@cr, rest::binary>>, @eb) do
+    if rest == "" do
+      true
+    else
+      :data_after_trailer
+    end
   end
 
-  defp do_recv(state, length, timeout) do
-    timeout_in_ms = System.convert_time_unit(timeout, :microsecond, :millisecond)
-    t1 = System.monotonic_time(:microsecond)
-    res = state.tcp.recv(state.socket, length, timeout_in_ms)
-    t2 = System.monotonic_time(:microsecond)
-    {res, t2 - t1}
+  defp trailer_check(packet, _last_buffer_byte) do
+    case :binary.match(packet, @trailer) do
+      :nomatch -> false
+      {pos, @trailer_length} when pos + @trailer_length == byte_size(packet) -> true
+      _ -> :data_after_trailer
+    end
   end
 
-  defp update_recv_time_credit(:infinity, _), do: {:infinity, 0}
+  defp update_receive_buffer(buffer, packet) do
+    case buffer do
```

Comment:
We don't need to case here. It should be fine to always `[buffer | packet]`

Reply:
The reason for having `buffer = []` case is to avoid creating a nested list with []. That is:
```elixir
iex(14)> buffer = []
[]       
iex(15)> reply = <<"abc">>
"abc"    
iex(16)> [buffer | reply] 
[[] | "abc"]
```

Reply:
Creating such a list should be fine :)

Reply:
Sure :-) but why do it if we can avoid it?
Probably a nitpick on my side :-)

Reply:
To avoid unecessary conditionals in the code. :)

Reply:
Right, so it probably comes down to whether it bugs you to see it if you're debugging, otherwise, they should cost the same amount of memory in words. 

```elixir
ex(1)> buffer = []
[]
iex(2)> buffer = [buffer | "a"]
[[] | "a"]
iex(3)> buffer = [buffer | "b"]
[[[] | "a"] | "b"]
iex(4)> :erts_debug.flat_size(buffer)
16
iex(5)> :erts_debug.size(buffer)
16
iex(6)> buffer1 = ["a"]
["a"]
iex(7)> buffer1 = [buffer1 | "b"]
[["a"] | "b"]
iex(8)> :erts_debug.flat_size(buffer1)
16
iex(9)> :erts_debug.flat_size(buffer)
16
```

No strong feelings here, just noticed 😄

Reply:
I would think that flattening nested lists into a binary would work better for lesser nested ones :-)
But yeah, let's go with less code instead. No strong feelings here as well :-)